Add public Boost headers explicitly

hebasto opened the pull request:
To check symbols in the code base, run:
```
git grep boost::multi_index::identity
git grep boost::multi_index::indexed_by
git grep boost::multi_index::tag
git grep boost::make_tuple
```

Hoping on the absence of conflicts with top-prio PRs :)

hebasto commented on src/node/miner.h:
```diff
@@ -14,7 +14,10 @@
+#include <boost/multi_index/indexed_by.hpp>
 #include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/tag.hpp>
 #include <boost/multi_index_container.hpp>
```
```
$ git grep -l boost::multi_index::tag
src/node/miner.h
src/txmempool.h
src/txrequest.cpp
```

The symbol is used in the header. IWYU agree :)

hebasto commented on test/lint/lint-includes.py:
```diff
@@ -21,17 +21,23 @@
                 ]
 
-EXPECTED_BOOST_INCLUDES = ["boost/date_time/posix_time/posix_time.hpp",
+EXPECTED_BOOST_INCLUDES = ["boost/assert.hpp",
```
I see no point at all. Is it OK to put this change into this PR ?

hebasto replied:
Or I can just revert the recent change?

theuni commented on test/lint/lint-includes.py:
```diff
@@ -21,17 +21,23 @@
+                           "boost/multi_index/indexed_by.hpp",
                            "boost/multi_index/ordered_index.hpp",
                            "boost/multi_index/sequenced_index.hpp",
+                           "boost/multi_index/tag.hpp",
```
I would expect to see `boost/multi_index_container.hpp` removed from this list as part of this PR, but it looks like it's still used:
```bash
 $ git grep multi_index_container.hpp origin/pr/27783
origin/pr/27783:node/miner.h:#include <boost/multi_index_container.hpp>
origin/pr/27783:txmempool.h:#include <boost/multi_index_container.hpp>
origin/pr/27783:txrequest.cpp:#include <boost/multi_index_container.hpp>
```

Is there some complication with those 3 that prevents us from using the specific headers rather than the umbrella one?

theuni replied:
Nevermind, I misremembered how `multi_index_container.hpp` worked. I was thinking it was just a dumb list of other includes, but it's more than that.

Still, does using `multi_index_container_fwd.hpp` help at all?

hebasto replied:
> Still, does using `multi_index_container_fwd.hpp` help at all?

No, it doesn't:
```
./txmempool.h:406:29: error: field ‘mapTx’ has incomplete type ‘CTxMemPool::indexed_transaction_set’ {aka ‘boost::multi_index::multi_index_container<CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::hashed_unique<mempoolentry_txid, SaltedTxidHasher>, boost::multi_index::hashed_unique<boost::multi_index::tag<index_by_wtxid>, mempoolentry_wtxid, SaltedTxidHasher>, boost::multi_index::ordered_non_unique<boost::multi_index::tag<descendant_score>, boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByDescendantScore>, boost::multi_index::ordered_non_unique<boost::multi_index::tag<entry_time>, boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByEntryTime>, boost::multi_index::ordered_non_unique<boost::multi_index::tag<ancestor_score>, boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByAncestorFee> > >’}
  406 |     indexed_transaction_set mapTx GUARDED_BY(cs);
      |   
```

hebasto committed:
test: Avoid `BOOST_ASSERT` macro

The `BOOST_ASSERT` macro is defined in the `boost/assert.hpp` header.
This change allows to skip `#include <boost/assert.hpp>`.

hebasto committed:
Add public Boost headers explicitly

hebasto commented:
Taken @MarcoFalke's [suggestion](https://github.com/bitcoin/bitcoin/pull/27783#discussion_r1211170115).

sedited reviewed (approved):
ACK 2484cacb7a6367b24e924dba0825c843b1dfc1c3
